SharedSource/UnitTests/Reflection/TypeExtensionsTest.cs

```diff
@@ -0,0 +1,170 @@
+// License, or (at your option) any later version.
+//
+// re-motion is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+//
+
+using System;
+using System.Reflection;
+using NUnit.Framework;
+using Remotion.Reflection;
+
+#nullable enable
+// ReSharper disable once CheckNamespace
+namespace Remotion.UnitTests.Reflection
+{
+  [TestFixture]
+  public class TypeExtensionsTest
+  {
+    [Test]
+    public void GetAssemblyQualifiedNameSafe_SystemType_AssemblyQualifiedName ()
+    {
+      var typeArgument = typeof (string);
+      const string expectedAssemblyQualifiedNamePrefix = "System.String,";
+
+      Assert.That (typeArgument.AssemblyQualifiedName, Is.Not.Null);
+      Assert.That (typeArgument.GetAssemblyQualifiedNameSafe(), Does.StartWith (expectedAssemblyQualifiedNamePrefix));
+    }
+
+    [Test]
+    public void GetAssemblyQualifiedNameSafe_GenericTypeParameter_TypeName ()
+    {
+      var genericTypeParameters = typeof (CustomType<>).GetTypeInfo().GenericTypeParameters;
+      const string expectedTypeArgumentName = "T";
+
+      Assert.That (genericTypeParameters, Has.Length.EqualTo (1));
+      Assert.That (genericTypeParameters[0].AssemblyQualifiedName, Is.Null);
+      Assert.That (genericTypeParameters[0].FullName, Is.Null);
+      Assert.That (genericTypeParameters[0].GetAssemblyQualifiedNameSafe(), Is.EqualTo (expectedTypeArgumentName));
+    }
+
+    [Test]
+    public void GetAssemblyQualifiedNameChecked_SystemType_AssemblyQualifiedName ()
+    {
+      var typeArgument = typeof (CustomType<string>).GetGenericArguments();
+      const string expectedAssemblyQualifiedNamePrefix = "System.String,";
+
+      Assert.That (typeArgument, Has.Length.EqualTo (1));
+      Assert.That (typeArgument[0].AssemblyQualifiedName, Is.Not.Null);
+      Assert.That (typeArgument[0].GetAssemblyQualifiedNameChecked(), Does.StartWith (expectedAssemblyQualifiedNamePrefix));
+    }
+
+    [Test]
+    public void GetAssemblyQualifiedNameChecked_GenericTypeParameter_Throws ()
+    {
+      var genericTypeArgumentsParameters = typeof (CustomType<>).GetTypeInfo().GenericTypeParameters;
+      const string expectedTypeArgumentName = "T";
+
+      Assert.That (genericTypeArgumentsParameters, Has.Length.EqualTo (1));
+      Assert.That (genericTypeArgumentsParameters[0].AssemblyQualifiedName, Is.Null);
+      Assert.That (
+          () => genericTypeArgumentsParameters[0].GetAssemblyQualifiedNameChecked(),
+          Throws.InvalidOperationException.With.Message.EqualTo (
+              string.Format ("Type '{0}' does not have an assembly qualified name.", expectedTypeArgumentName)));
```

Comment:
```suggestion
          Throws.InvalidOperationException.With.Message.EqualTo ("Type 'T' does not have an assembly qualified name."));
```
In cases such as this, I usually avoid string.format, opting for a completed string instead. The recommendation is always to have the expectations with as little logic as possible. string.format is only chosen if the actual value either changes with each test run or is otherwise highly volatile or dependent on external dependencies (e.g. localized strings that get updated regularily or are reused a lot and would thus result in many updates if changed). It also makes the test easier to read.

Reply:
I think this was also partly elaborated in the last meeting. My intention behind the layout with the constant and the `string.format` was just that you see - especially for when the value is not volatile or some other developer might stumble upon the test at a later point in time - at one glance which parts of the expectation can (in some way) influenced by the test author (in this case the (generic) type with `CustomType`). So you would see the metaphoric screws to adjust, even without making oneself too familiar with the intrinsics of the test.

But, of course, one has gauge whether that's worth it or how one would want to value that. In any way, your suggestions make the test undeniably clearer, more easily comprehensible and more concise. I have updated the code accordingly.

Reply:
Thanks. And yes, there's this dichotomy between the SUT code and test code. The SUT must be clean in terms of duplication, obvious in the way the result is constructed, etc. The test must be obvious in terms of the expectation and obviousness as to where the error is. Once that is achieved, fixing the string shouldn't be difficult in the event that ever changes. 

But yes, if we needed different types based on the compilation target, etc, things would be very different indeed.